narwhal/primary/src/primary.rs

```diff
@@ -898,6 +890,81 @@ impl PrimaryReceiverHandler {
             missing: Vec::new(),
         })
     }
+
+    // Tries to accept certificates if they have been requested from the header author.
+    // The filtering is to avoid overload from unrequested certificates. It is ok that this
+    // filter may result in a certificate never arriving via header proposals, because
+    // liveness is guaranteed by certificate fetching.
+    async fn try_accept_unknown_parents(
+        &self,
+        header: &Header,
+        mut parents: Vec<Certificate>,
+    ) -> DagResult<()> {
+        {
+            let parent_digests = self.parent_digests.lock();
+            parents.retain(|cert| {
+                let Some(from) = parent_digests.get(&(cert.round(), cert.digest())) else {
+                    return false;
+                };
+                // Only process a certificate from the primary where it is first known.
+                *from == header.author()
+            });
+        }
+        for parent in parents {
+            self.synchronizer.try_accept_certificate(parent).await?;
+        }
+        Ok(())
+    }
+
+    /// Gets parent certificate digests not known before, in storage, among suspended certificates,
+    /// or being requested from other header proposers.
+    async fn get_unknown_parent_digests(
+        &self,
+        header: &Header,
+    ) -> DagResult<Vec<CertificateDigest>> {
+        // Get digests not known by the synchronizer, in storage or among suspended certificates.
+        let mut digests = self.synchronizer.get_unknown_parent_digests(header).await?;
+
+        // Maximum header age is chosen to strike a balance between allowing for slightly older
+        // certificates to still have a chance to be included in the DAG while not wasting
+        // resources on very old vote requests. This value affects performance but not correctness
+        // of the algorithm.
+        const HEADER_AGE_LIMIT: Round = 3;
```

Comment:
The header age limit used to be GC round. But the argument is that if most primaries are at round R+1 already, a new certificate a round R would unlikely become a parent to other certificates. Increasing the limit may be mostly harmless, but in theory it can increase the time uncommittable batches get retried in a new header. We can definitely spend time to experiment. In future with chained certificates, this limit will need to be relaxed.